arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedPageIterator.java

```diff
@@ -0,0 +1,444 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.arrow.vectorized.parquet;
+
+import java.io.IOException;
+import org.apache.arrow.vector.DecimalVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.VarBinaryVector;
+import org.apache.iceberg.arrow.vectorized.NullabilityHolder;
+import org.apache.iceberg.parquet.BasePageIterator;
+import org.apache.iceberg.parquet.ValuesAsBytesReader;
+import org.apache.parquet.CorruptDeltaByteArrays;
+import org.apache.parquet.bytes.ByteBufferInputStream;
+import org.apache.parquet.bytes.BytesUtils;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.column.Dictionary;
+import org.apache.parquet.column.Encoding;
+import org.apache.parquet.column.values.RequiresPreviousReader;
+import org.apache.parquet.column.values.ValuesReader;
+import org.apache.parquet.io.ParquetDecodingException;
+
+public class VectorizedPageIterator extends BasePageIterator {
+  private final boolean setArrowValidityVector;
+
+  public VectorizedPageIterator(ColumnDescriptor desc, String writerVersion, boolean setValidityVector) {
+    super(desc, writerVersion);
+    this.setArrowValidityVector = setValidityVector;
+  }
+
+  private boolean eagerDecodeDictionary;
+  private ValuesAsBytesReader plainValuesReader = null;
+  private VectorizedDictionaryEncodedParquetValuesReader dictionaryEncodedValuesReader = null;
+  private boolean allPagesDictEncoded;
+
+  // Dictionary is set per row group
+  public void setDictionaryForColumn(Dictionary dict, boolean allDictEncoded) {
+    this.dictionary = dict;
+    this.allPagesDictEncoded = allDictEncoded;
+  }
+
+  @Override
+  protected void reset() {
+    this.page = null;
+    this.triplesCount = 0;
+    this.triplesRead = 0;
+    this.repetitionLevels = null;
+    this.plainValuesReader = null;
+    this.vectorizedDefinitionLevelReader = null;
+    this.hasNext = false;
+  }
+
+  public int currentPageCount() {
+    return triplesCount;
+  }
+
+  public boolean hasNext() {
+    return hasNext;
+  }
+
+  /**
+   * Method for reading a batch of dictionary ids from the dicitonary encoded data pages. Like definition levels,
+   * dictionary ids in Parquet are RLE/bin-packed encoded as well.
+   */
+  public int nextBatchDictionaryIds(
+      final IntVector vector, final int expectedBatchSize,
+      final int numValsInVector,
+      NullabilityHolder holder) {
+    final int actualBatchSize = getActualBatchSize(expectedBatchSize);
+    if (actualBatchSize <= 0) {
+      return 0;
+    }
+    ((VectorizedParquetValuesReader) vectorizedDefinitionLevelReader).readBatchOfDictionaryIds(
```

Comment:
It seems strange to me that this is always cast to a `VectorizedParquetValuesReader` in this class. Why not make the field that type?

Reply:
There is a dependency issue here. VectorizedParquetValuesReader is in arrow module where as the base class BasePageIterator is in parquet module. The arrow module depends on parquet. So I cannot declare the type of vectorizedDefinitionLevelReader in BasePageIterator to VectorizedParquetValuesReader as it would result in a circular dependency.